src/main/java/de/tum/cit/aet/artemis/lecture/web/LectureResource.java

```diff
@@ -201,18 +213,75 @@ public ResponseEntity<Set<Lecture>> getLecturesForCourse(@PathVariable Long cour
      */
     @GetMapping("courses/{courseId}/lectures-with-slides")
     @EnforceAtLeastStudent
-    public ResponseEntity<Set<Lecture>> getLecturesWithSlidesForCourse(@PathVariable Long courseId) {
-        log.debug("REST request to get all Lectures with slides of the units for the course with id : {}", courseId);
+    public ResponseEntity<List<LectureDTO>> getLecturesWithSlidesForCourse(@PathVariable Long courseId) {
+        log.info("Getting all lectures with slides for course {}", courseId);
+        long start = System.currentTimeMillis();
+        var course = courseRepository.findByIdElseThrow(courseId);
+        var user = userRepository.getUserWithGroupsAndAuthorities();
+        authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);
+
+        var lectures = lectureRepository.findAllByCourseIdWithAttachmentsAndLectureUnits(courseId).stream().filter(Lecture::isVisibleToStudents).collect(Collectors.toSet());
+        Set<Long> attachmentUnitIds = lectures.stream().flatMap(lecture -> lecture.getLectureUnits().stream()).filter(lectureUnit -> lectureUnit instanceof AttachmentUnit)
+                .map(DomainObject::getId).collect(Collectors.toSet());
+
+        // Load slides separately to avoid too large data exchange
+        Set<SlideDTO> slides = slideRepository.findVisibleSlidesByAttachmentUnits(attachmentUnitIds);
+
+        // Group slides by attachment unit id to combine them into the DTOs
+        Map<Long, List<SlideDTO>> slidesByAttachmentUnitId = slides.stream().collect(Collectors.groupingBy(SlideDTO::attachmentUnitId));
+        // Convert visible lectures to DTOs (filtering active attachments) and add non hidden slides to the DTOs
+        List<LectureDTO> lectureDTOs = lectures.stream().map(LectureDTO::from).peek(lectureDTO -> {
+            List<AttachmentUnitDTO> attachmentUnitDTOs = lectureDTO.lectureUnits;
+            for (AttachmentUnitDTO attachmentUnitDTO : attachmentUnitDTOs) {
+                List<SlideDTO> slidesForAttachmentUnit = slidesByAttachmentUnitId.get(attachmentUnitDTO.id);
+                if (slidesForAttachmentUnit != null) {
+                    // remove unnecessary fields from the slide DTOs
+                    var finalSlides = slidesForAttachmentUnit.stream().map(slideDTO -> new SlideDTO(slideDTO.id(), slideDTO.slideNumber(), null, null))
+                            .sorted(Comparator.comparingInt(SlideDTO::slideNumber)).toList();
+                    attachmentUnitDTO.slides.addAll(finalSlides);
+                }
+            }
+        }).sorted(Comparator.comparingLong(LectureDTO::id)).toList();
```

Comment:
I suggest separating the side effect in an easy case like this

```suggestion
        List<LectureDTO> lectureDTOs = lectures.stream()
            .map(LectureDTO::from)
            .sorted(Comparator.comparingLong(LectureDTO::id))
            .toList();
        
        lectureDTOs.forEach(lectureDTO -> {
            for (AttachmentUnitDTO attachmentUnitDTO : lectureDTO.lectureUnits) {
                List<SlideDTO> slidesForAttachmentUnit = slidesByAttachmentUnitId.get(attachmentUnitDTO.id);
                if (slidesForAttachmentUnit != null) {
                    // remove unnecessary fields from the slide DTOs
                    var finalSlides = slidesForAttachmentUnit.stream()
                        .map(slideDTO -> new SlideDTO(slideDTO.id(), slideDTO.slideNumber(), null, null))
                        .sorted(Comparator.comparingInt(SlideDTO::slideNumber))
                        .toList();
                    attachmentUnitDTO.slides.addAll(finalSlides);
                }
            }
        });
```